[INF-0002] Fill out sections on Validations. (#235)
microsoft/hlsl-specs: proposals/infra/INF-0002-root-signature-in-clang.md

    RootSRV : 'SRV' '(' tReg (',' 'space' '=' NUMBER)?
          (',' 'visibility' '=' SHADER_VISIBILITY)?
          (',' 'flags' '=' DATA_FLAGS)? ')'

    RootUAV : 'UAV' '(' uReg (',' 'space' '=' NUMBER)?
          (',' 'visibility' '=' SHADER_VISIBILITY)?
          (',' 'flags' '=' DATA_FLAGS)? ')'

    DescriptorTable : 'DescriptorTable' '(' (DTClause(|DTClause)?)?
          (',' 'visibility' '=' SHADER_VISIBILITY)? ')'

    DTClause : CBV | SRV | UAV | Sampler

    CBV : 'CBV' '(' bReg (',' 'numDescriptors' '=' NUMBER)?
          (',' 'space' '=' NUMBER)?
          (',' 'offset' '=' DESCRIPTOR_RANGE_OFFSET)?
          (',' 'flags' '=' DATA_FLAGS)? ')'

    SRV : 'SRV' '(' tReg (',' 'numDescriptors' '=' NUMBER)?
          (',' 'space' '=' NUMBER)?
          (',' 'offset' '=' DESCRIPTOR_RANGE_OFFSET)?
          (',' 'flags' '=' DATA_FLAGS)? ')'

    UAV : 'UAV' '(' uReg (',' 'numDescriptors' '=' NUMBER)?
          (',' 'space' '=' NUMBER)?
          (',' 'offset' '=' DESCRIPTOR_RANGE_OFFSET)?
          (',' 'flags' '=' DATA_FLAGS)? ')'

    Sampler : 'Sampler' '(' sReg (',' 'numDescriptors' '=' NUMBER)?
          (',' 'space' '=' NUMBER)?

>           (',' 'offset' '=' DESCRIPTOR_RANGE_OFFSET)? (',' 'flags' '=' NUMBER)? ')'

Reviewer: Should this actually be `'flags' '=' DATA_FLAGS`?

Author: Fixed.

Reviewer: This doesn't seem to have been changed.

Author: Should be updated now.

>           (',' 'offset' '=' DESCRIPTOR_RANGE_OFFSET)? (',' 'flags' '=' DATA_FLAGS)? ')'

Reviewer: I'm having a hard time getting Samplers to work in DXC:

https://godbolt.org/z/GdKsK7TbY
https://godbolt.org/z/9qf1GE6os

Author: The special rule for Sampler is only following works.
- None
- DescriptorsVolatile
- DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS
We can put that in grammar if you like.

Author: Put this into grammar.

    SHADER_VISIBILITY : 'SHADER_VISIBILITY_ALL' | 'SHADER_VISIBILITY_VERTEX' |
                        'SHADER_VISIBILITY_HULL' |
                        'SHADER_VISIBILITY_DOMAIN' |
                        'SHADER_VISIBILITY_GEOMETRY' |
                        'SHADER_VISIBILITY_PIXEL' |
                        'SHADER_VISIBILITY_AMPLIFICATION' |
                        'SHADER_VISIBILITY_MESH'

>     DATA_FLAGS : 'DATA_STATIC_WHILE_SET_AT_EXECUTE' | 'DATA_VOLATILE'

Reviewer: It looks like `0` is also a valid `DATA_FLAG`: https://godbolt.org/z/MdeK1o6j5

Author: Updated.

every language that targets DirectX would want to adopt. For this reason we need
to ensure that our solution doesn't unnecessarily tie the non-HLSL parts to it.

> ### Root Signature Grammar

Reviewer 2: Is there are public reference we could link to for the grammar?

Author: Cannot find reference. And most things are for fxc which doesn't include new flags added for newer shader models.

    RootFlags : 'RootFlags' '(' (RootFlag(|RootFlag)?)? ')'

    RootFlag : 0 |
               'ALLOW_INPUT_ASSEMBLER_INPUT_LAYOUT' |
               'DENY_VERTEX_SHADER_ROOT_ACCESS' |
               'DENY_HULL_SHADER_ROOT_ACCESS' |
               'DENY_DOMAIN_SHADER_ROOT_ACCESS' |
               'DENY_GEOMETRY_SHADER_ROOT_ACCESS' |
               'DENY_PIXEL_SHADER_ROOT_ACCESS' |
               'DENY_AMPLIFICATION_SHADER_ROOT_ACCESS' |
               'DENY_MESH_SHADER_ROOT_ACCESS' |
               'ALLOW_STREAM_OUTPUT' |
               'LOCAL_ROOT_SIGNATURE' |
               'CBV_SRV_UAV_HEAP_DIRECTLY_INDEXED' |
               'SAMPLER_HEAP_DIRECTLY_INDEXED' |

>                'AllowLowTierReservedHwCbLimit'

Reviewer: > AllowLowTierReservedHwCbLimit

What is this flag?  And why does not look like any of the others?

    RootCBV : 'CBV' '(' bReg (',' 'space' '=' NUMBER)?
          (',' 'visibility' '=' SHADER_VISIBILITY)?
          (',' 'flags' '=' ROOT_DESCRIPTOR_FLAGS)? ')'

    RootSRV : 'SRV' '(' tReg (',' 'space' '=' NUMBER)?
          (',' 'visibility' '=' SHADER_VISIBILITY)?
          (',' 'flags' '=' ROOT_DESCRIPTOR_FLAGS)? ')'

    RootUAV : 'UAV' '(' uReg (',' 'space' '=' NUMBER)?
          (',' 'visibility' '=' SHADER_VISIBILITY)?
          (',' 'flags' '=' ROOT_DESCRIPTOR_FLAGS)? ')'

    DescriptorTable : 'DescriptorTable' '(' (DTClause(|DTClause)?)?
          (',' 'visibility' '=' SHADER_VISIBILITY)? ')'

    DTClause : CBV | SRV | UAV | Sampler

    DESCRIPTOR_RANGE_FLAGS :
        0 |
        'DESCRIPTORS_VOLATILE' |
        'DATA_VOLATILE' |
        'DATA_STATIC' |
        'DATA_STATIC_WHILE_SET_AT_EXECUTE' |
        'DESCRIPTORS_VOLATILE' '|' 'DATA_VOLATILE' |
        'DESCRIPTORS_VOLATILE' '|' 'DATA_STATIC_WHILE_SET_AT_EXECUTE' |
        'DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS' |
        'DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS' '|' 'DATA_VOLATILE' |
        'DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS' '|' 'DATA_STATIC' |

>         'DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS' '|' 'DATA_STATIC_WHILE_SET_AT_EXECUTE'

Reviewer: This doesn't look like a grammar to me.  I think you're trying to list possible valid combinations of flags, with one valid combination per line?  But I don't think you're saying that we'd reject `DATA_VOLATILE | DESCRIPTORS_VOLATILE`?  This seems like a mix of formal and informal notation here.

I think you should keep it simple - have the grammar able to capture all the values of the `D3D12_DESCRIPTOR_RANGE_FLAG` enum, and then treat validating the flag combinations as part of semantic analysis.  The API doesn't have a distinction between sampler range flags and descriptor range flags, I don't think that the grammar should either.

Author: Updated.

### Validations in Sema

#### All the values should be legal.

>   ShaderVisibility

Author: Updated.

      'SHADER_VISIBILITY_ALL'
      'SHADER_VISIBILITY_VERTEX'
      'SHADER_VISIBILITY_HULL'
      'SHADER_VISIBILITY_DOMAIN'
      'SHADER_VISIBILITY_GEOMETRY'
      'SHADER_VISIBILITY_PIXEL'
      'SHADER_VISIBILITY_AMPLIFICATION'

>       'SHADER_VISIBILITY_MESH'

Reviewer: There's no way to specify an invalid shader visibility, is there?  I think some of these should fall under "it must be syntactically correct".  You found a grammar somewhere for the root signature DSL?  Maybe we can include that in this section, or link to it?

Author: Added grammar.

  ParameterType

      CBV
      SRV
      UAV
      DescriptorTable
      Constants32Bit

  RootDescriptorFlags

      None
      DataVolatile
      DataStaticWihleSetAtExecute
      DataStatic

  DescriptorRangeFlags

    For Sampler
      None
      DescriptorsVolatile
      DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS

>     For resource

Reviewer: Is this saying that not all combinations of flags for resources are valid?  If so, then this should also be validated in codegen.

Reviewer: In fact, I think that in any cases where the metadata allows invalid root signatures to be represented we need to validate them again during codegen.  Codegen cannot assume that the metadata it has been given is valid.

Author: Yes. Many of these value checks are covered by syntax check in Sema, 
list all of these for DXIL generation check to reference.

      None
      DESCRIPTORS_VOLATILE,
      DATA_VOLATILE,
      DATA_STATIC,
      DATA_STATIC_WHILE_SET_AT_EXECUTE,
      DESCRIPTORS_VOLATILE | DATA_VOLATILE,
      DESCRIPTORS_VOLATILE | DATA_STATIC_WHILE_SET_AT_EXECUTE,

      DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS,
      DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS | DATA_VOLATILE,
      DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS | DATA_STATIC,
      DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS | DATA_STATIC_WHILE_SET_AT_EXECUTE,

  StaticSampler

      Max/MinLOD cannot be NaN.
      Comparison filter must have ComparisonFunc not equal to 0.
      MaxAnisotropy cannot exceed 16.
      MipLODBias must be within range of [-16, 15.99].

  Space

      The range 0xFFFFFFF0 to 0xFFFFFFFF is reserved.

>       "CBV(b0, space=4294967295)" is invalid duo to the use of reserved space 0xFFFFFFFF.

Reviewer: ```suggestion
      "CBV(b0, space=4294967295)" is invalid due to the use of reserved space 0xFFFFFFFF.
```

Author: Fixed.

#### Resource ranges must not overlap.

      "CBV(b2), DescriptorTable(CBV(b0, numDescriptors=5))" will result in an 
      error due to overlapping at b2.

### Metadata Schema

Here is the metadata presentation for the root signature.

Root Signature
| | Version | Flag | RootParameters | StaticSamplers |
|-| --- | --- | --- | --- |

> | | i32 | i32 | list of RootParameters (RootConstant, RootDescriptor, DescriptorTable) | list of StaticSamplers |

Reviewer: This doesn't match the format that was proposed previously.  See the "Root Signatures in the LLVM IR" section above.

Also, note that there's a separate issue for updating the metadata schema.  I suggest you tackle the validation and metadata changes in different PRs.

Author: Removed.

| | i32 |  i32 | i32 | i32 | i32 |

DescriptorTable
| | ParameterType | Visibility | DescriptorRanges
|-| --- | --- | --- |
|  | i32 |  i32 | list of DescriptorRanges |

DescriptorRange
| | RangeType | NumDescriptors | BaseShaderRegister | Space | Flags | Offset |
|-| --- | --- | --- | --- | --- | --- |
| | i32 | i32 | i32 | i32 | i32 | i32 |

StaticSampler

|| Filter | AddressU | AddressV | AddressW | MipLODBias | MaxAnisotropy | ComparisonFunc | BoarderColor | MinLOD | MaxLOD | Register | Space | Visibility |
| - | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| | i32 | i32 | i32 | i32 | float | i32 | i32 | i32 | float | float | i32 | i32 | i32 |

### Validations during Codegen

#### All the things validated in Sema.

#### Resource used in DXIL must be fully bound in root signature.
```
  // B is bound to t1, but not in root descriptor.
  Buffer<float> B : register(t1);
  [RootSignature("")]
  void main() : SV_Target {
    return B[0];
  }
```

#### Root Signature Flag must match DXIL.
```
  // Used dynamic resource but missing CBVSRVUAVHeapDirectlyIndexed flag.
  [RootSignature("")]
  void main() : SV_Target {

>     Buffer<float> B = ResourceDescriptorHead[0];

Reviewer: ```suggestion
    Buffer<float> B = ResourceDescriptorHeap[0];
```

Author: Fixed.

### Validations in Sema

#### All the values should be legal.

Some of these values are covered by syntactical checks in Sema.

List all here for reference during DXIL generation checks.

##### ShaderVisibility
- SHADER_VISIBILITY_ALL
- SHADER_VISIBILITY_ALL
- SHADER_VISIBILITY_VERTEX
- SHADER_VISIBILITY_HULL
- SHADER_VISIBILITY_DOMAIN
- SHADER_VISIBILITY_GEOMETRY
- SHADER_VISIBILITY_PIXEL
- SHADER_VISIBILITY_AMPLIFICATION
- SHADER_VISIBILITY_MESH

##### ParameterType
- CBV
- SRV
- UAV
- DescriptorTable

> - Constants32Bit

Reviewer: The terms used here don't match the terms used in the grammar.  I'm still unconvinced that there is any value at all in listing anything here that is just "can be parsed by the grammar".  The more interesting validations to list are the ones that are syntactically correct but have semantic errors.

Author: Fixed.

##### RootDescriptorFlags
- None
- DataVolatile
- DataStaticWihleSetAtExecute
- DataStatic

##### DescriptorRangeFlags

###### Sampler
- None
- DescriptorsVolatile
- DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS

###### Resource
- None
- DESCRIPTORS_VOLATILE,
- DATA_VOLATILE,
- DATA_STATIC,
- DATA_STATIC_WHILE_SET_AT_EXECUTE,
- DESCRIPTORS_VOLATILE | DATA_VOLATILE,
- DESCRIPTORS_VOLATILE | DATA_STATIC_WHILE_SET_AT_EXECUTE,
- DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS,
- DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS | DATA_VOLATILE,
- DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS | DATA_STATIC,
- DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS | DATA_STATIC_WHILE_SET_AT_EXECUTE,

##### StaticSampler
- Max/MinLOD cannot be NaN.

> - Comparison filter must have ComparisonFunc not equal to 0.

Reviewer: The grammar has a list of valid ComparisonFuncs, so how would a user specify a syntactically correct StaticSampler is a ComparsionFunc equal to 0?

Author: For the backend, the ComparisonFunc is just an i32 which could be 0.

Reviewer: This section is describing the validations performed in Sema.  I can't see any way in Sema for this to happen.  There's a separate section for describing the validations performed during codegen.

Author: Moved to DXIL generation part.

### Validations in Sema

#### All the values should be legal.

Most values like ShaderVisibility/ParameterType are covered by syntactical checks in Sema.
Only list special rule here.

##### DescriptorRangeFlags

###### Sampler
- None
- DescriptorsVolatile

> - DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS

Reviewer: I can't figure out what any of these mean.  Can you make it super clear and explicit what the is being validated here? The only references to DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS I can find are in this section on validation.  What does it mean? Is this meant to a supported flag somewhere?

What is "DescriptorsVolatile"?

Author: Here's doc for [DescriptorsVolatile](https://learn.microsoft.com/en-us/windows/win32/direct3d12/root-signature-version-1-1#descriptors_volatile) .
Cannot find doc about DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS as well :(
But from the last part of the [DescriptorsVolatile](https://learn.microsoft.com/en-us/windows/win32/direct3d12/root-signature-version-1-1#descriptors_volatile), it might be used when depend on safe out of bounds memory access behavior when accessing descriptors, but don't want to set DescriptorsVolatile.

###### Resource
- None
- DESCRIPTORS_VOLATILE,
- DATA_VOLATILE,
- DATA_STATIC,
- DATA_STATIC_WHILE_SET_AT_EXECUTE,
- DESCRIPTORS_VOLATILE | DATA_VOLATILE,
- DESCRIPTORS_VOLATILE | DATA_STATIC_WHILE_SET_AT_EXECUTE,
- DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS,
- DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS | DATA_VOLATILE,
- DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS | DATA_STATIC,
- DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS | DATA_STATIC_WHILE_SET_AT_EXECUTE,

##### StaticSampler
- Max/MinLOD cannot be NaN.
- MaxAnisotropy cannot exceed 16.
- MipLODBias must be within range of [-16, 15.99].

##### Register Space

  The range 0xFFFFFFF0 to 0xFFFFFFFF is reserved.
```
  "CBV(b0, space=4294967295)" is invalid due to the use of reserved space 0xFFFFFFFF.
```

#### Resource ranges must not overlap.
```
  "CBV(b2), DescriptorTable(CBV(b0, numDescriptors=5))" will result in an error
  due to overlapping at b2.
```

### Metadata Schema

TODO

### Validations during DXIL generation
  [Validations In Sema](#validations-in-sema)
#### All the things validated in Sema.

>   In DXIL generation, values like ShaderVisibility/ParameterType need to be

Reviewer: Let's just list all of the things "like ShaderVisibility/ParameterType" here.

Author: Updated.

  - 0
  - DESCRIPTORS_VOLATILE
  - DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS

- For DESCRIPTOR_RANGE_FLAGS on a CBV/SRV/UAV, only the following values are
   valid
  - 0
  - DESCRIPTORS_VOLATILE
  - DATA_VOLATILE
  - DATA_STATIC
  - DATA_STATIC_WHILE_SET_AT_EXECUTE
  - DESCRIPTORS_VOLATILE | DATA_VOLATILE
  - DESCRIPTORS_VOLATILE | DATA_STATIC_WHILE_SET_AT_EXECUTE
  - DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS
  - DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS | DATA_VOLATILE
  - DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS | DATA_STATIC
  - DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS | DATA_STATIC_WHILE_SET_AT_EXECUTE

- StaticSampler
  - Max/MinLOD cannot be NaN.
  - MaxAnisotropy cannot exceed 16.
  - MipLODBias must be within range of [-16, 15.99].

- Register Space
  -The range 0xFFFFFFF0 to 0xFFFFFFFF is reserved.
  ```
  "CBV(b0, space=4294967295)" is invalid due to the use of reserved space 0xFFFFFFFF.
  ```

- Resource ranges must not overlap.
  ```
  "CBV(b2), DescriptorTable(CBV(b0, numDescriptors=5))" will result in an error
  due to overlapping at b2.
  ```

### Metadata Schema

TODO

### Validations during DXIL generation

>   [Validations In Sema](#validations-in-sema)

Reviewer: So there's a link to the "Validations In Sema" section, and then there's a subheading following it call "All the things validated in Sema".  Can you clean this up please to make it easier to read?

Author: Updated.

#### All the things validated in Sema.

>   In DXIL generation, values like ShaderVisibility/ParameterType need to be
>   checked to make sure they are in correct range.

Reviewer: ```suggestion
  In DXIL generation, various values need to be
  checked to make sure they are in correct range:
```

Author: Updated.

- Valid values for ShaderVisibility
  - SHADER_VISIBILITY_ALL
  - SHADER_VISIBILITY_VERTEX
  - SHADER_VISIBILITY_HULL
  - SHADER_VISIBILITY_DOMAIN
  - SHADER_VISIBILITY_GEOMETRY
  - SHADER_VISIBILITY_PIXEL
  - SHADER_VISIBILITY_AMPLIFICATION
  - SHADER_VISIBILITY_MESH

- Valid values for ParameterType
  - CBV
  - SRV
  - UAV
  - DescriptorTable

>   - Constants32Bit

Reviewer: In the metadata we have entries with names like:

* RootFlags
* RootCBV
* StaticSampler
* DescriptorTable

Maybe this'll be something that needs revisiting once the metadata format is fleshed out,

Author: The ParameterType here is only for the Parameter part of the root signature.
For the entries in metatdata, they're for the whole root signature, not the same thing.
We could visit metadata entries when fill the metadata section.
The parameter type here should be correct.

Reviewer: At this point, I think we're validating that the values provided by the metadata are valid - so that if errors are encountered we can refer to them in terms of the metadata.

What, specifically, are you envisioning "ParameterType" is actually validating here? If it is the value of D3D12_ROOT_PARAMETER::ParameterType, then this is something that is generated by codegen - we shouldn't need to do a second pass of validation on the output of codegen. What we need to validate is the _input_ to codegen, which is the metadata.

Author: I see. Remove ParameterType.

Reviewer: I've added a note to #234 about ensuring that the validation section includes information about what needs to be validated for the metadata.

    - FILTER_MAXIMUM_MIN_MAG_MIP_POINT
    - FILTER_MAXIMUM_MIN_MAG_POINT_MIP_LINEAR
    - FILTER_MAXIMUM_MIN_POINT_MAG_LINEAR_MIP_POINT
    - FILTER_MAXIMUM_MIN_POINT_MAG_MIP_LINEAR
    - FILTER_MAXIMUM_MIN_LINEAR_MAG_MIP_POINT
    - FILTER_MAXIMUM_MIN_LINEAR_MAG_POINT_MIP_LINEAR
    - FILTER_MAXIMUM_MIN_MAG_LINEAR_MIP_POINT
    - FILTER_MAXIMUM_MIN_MAG_MIP_LINEAR
    - FILTER_MAXIMUM_ANISOTROPIC

  - Valid values for TextureAddress
    - TEXTURE_ADDRESS_WRAP 
    - TEXTURE_ADDRESS_MIRROR  
    - TEXTURE_ADDRESS_CLAMP 
    - TEXTURE_ADDRESS_BORDER 
    - TEXTURE_ADDRESS_MIRROR_ONCE

  - Valid values for ComparisonFunc
    - 0
    - COMPARISON_NEVER 
    - COMPARISON_LESS 
    - COMPARISON_EQUAL 
    - COMPARISON_LESS_EQUAL 
    - COMPARISON_GREATER 
    - COMPARISON_NOT_EQUAL 
    - COMPARISON_GREATER_EQUAL 
    - COMPARISON_ALWAYS

  - Valid values for StaticBorderColor
    - STATIC_BORDER_COLOR_TRANSPARENT_BLACK 
    - STATIC_BORDER_COLOR_OPAQUE_BLACK 
    - STATIC_BORDER_COLOR_OPAQUE_WHITE

  - Comparison filter must have ComparisonFunc not equal to 0.
      ```
      When the Filter of a StaticSampler is FILTER_COMPARISON*,
      the ComparisonFunc cannot be 0.

>       ```

Reviewer: Should the `RootFlag`s also be validated?

Author: Updated.

#### Resource used in DXIL must be fully bound in root signature.
```
  // B is bound to t1, but not in root descriptor.
  Buffer<float> B : register(t1);
  [RootSignature("")]
  void main() : SV_Target {
    return B[0];
  }
```

#### Root Signature Flag must match DXIL.
```

>   // Used dynamic resource but missing CBVSRVUAVHeapDirectlyIndexed flag.

Reviewer: > CBVSRVUAVHeapDirectlyIndexed flag

I think that this is the first reference to this flag in this document.  I think it is a root flag?  It doesn't appear in the list of root flags in the grammar.

Author: Good catch, missed many root flags :(.
Updated.

Most values like ShaderVisibility/ParameterType are covered by syntactical checks in Sema.
Only list special rule here.

> - DESCRIPTOR_RANGE_FLAGS for Sampler

Reviewer: I think that what you're trying to say is during validation, descriptor ranges flags will be checked, and here is the list of valid flags for Sampler descriptor ranges, and the valid flags for UAV, CBV and SRC descriptor ranges.

These specs / proposals need to be written so that they can be read by someone who hasn't had any of the conversations and design meetings and reviews of existing code that you've had, so it is not enough to write them in note form.

Author: Updated.
Hope it make things clearer.

### Validations in Sema

#### All the values should be legal.

Most values like ShaderVisibility/ParameterType are covered by syntactical 
checks in Sema.

> Only list special rule here.

Author: Added in https://github.com/microsoft/hlsl-specs/pull/238

    - FILTER_MAXIMUM_MIN_POINT_MAG_LINEAR_MIP_POINT
    - FILTER_MAXIMUM_MIN_POINT_MAG_MIP_LINEAR
    - FILTER_MAXIMUM_MIN_LINEAR_MAG_MIP_POINT
    - FILTER_MAXIMUM_MIN_LINEAR_MAG_POINT_MIP_LINEAR
    - FILTER_MAXIMUM_MIN_MAG_LINEAR_MIP_POINT
    - FILTER_MAXIMUM_MIN_MAG_MIP_LINEAR
    - FILTER_MAXIMUM_ANISOTROPIC

  - Valid values for TextureAddress
    - TEXTURE_ADDRESS_WRAP 
    - TEXTURE_ADDRESS_MIRROR  
    - TEXTURE_ADDRESS_CLAMP 
    - TEXTURE_ADDRESS_BORDER 
    - TEXTURE_ADDRESS_MIRROR_ONCE

  - Valid values for ComparisonFunc
    - 0
    - COMPARISON_NEVER 
    - COMPARISON_LESS 
    - COMPARISON_EQUAL 
    - COMPARISON_LESS_EQUAL 
    - COMPARISON_GREATER 
    - COMPARISON_NOT_EQUAL 
    - COMPARISON_GREATER_EQUAL 
    - COMPARISON_ALWAYS

  - Valid values for StaticBorderColor
    - STATIC_BORDER_COLOR_TRANSPARENT_BLACK 
    - STATIC_BORDER_COLOR_OPAQUE_BLACK 
    - STATIC_BORDER_COLOR_OPAQUE_WHITE

  - Comparison filter must have ComparisonFunc not equal to 0.
      ```
      When the Filter of a StaticSampler is FILTER_COMPARISON*,
      the ComparisonFunc cannot be 0.
      ```

#### Resource used in DXIL must be fully bound in root signature.
```

>   // B is bound to t1, but not in root descriptor.

Reviewer 2: Does it need to be bound as a root descriptor? Or it only needs to be bound in the root signature?

Author: Good catch.
Fixed with https://github.com/microsoft/hlsl-specs/pull/238

  In DXIL generation, values like ShaderVisibility/ParameterType need to be
  checked to make sure they are in correct range.

- ShaderVisibility

>   - SHADER_VISIBILITY_ALL

Reviewer 3: SHADER_VISIBLITY_ALL is listed twice here.

Author: Fixed.
